### PR TITLE
[Form] fix ViolationMapper was always generating a localized label for each FormType

### PR DESCRIPTION
**Pull request by romaricdrigon**
| Q             | A
| ------------- | ---
| Branch?       | 5.2 (fix new behavior from 5.2)
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | /
| License       | MIT
| Doc PR        | /

**Explanation of the issue**

In Symfony 5.2, the `{{ label }}` placeholder can be used in constraint messages (-> introduced in commit 0d9f44235c3c6b4cfeb3e4edf28704b6a255f403)

However, the way it was coded is introducing a small side effect: now, every time there is validation error, `ViolationMapper` will ask the Form Type its `label`, and if not `false`, it will try to translate it.

**Why it is important/why it causes a BC break**

Since by default `AbstractType` does not have any `label`, it also introduces a minor BC break.  
I will explain it with an example: in a project I work on, we check we don't have any missing translation. Sometimes we have violation errors bound to form ; then current code will get Form `label`, which in `null` in form type classes (which is quite usual I believe), so it will generate one, and pass that one to translator. And we see a lot on erroneous missing translations.

**Proposed fix**

This fix moves all this logic into a `if`, so `ViolationMapper` call the translator component only if `{{ label }}` placeholder is used in constraint error message. 
On top of fixing BC, it has the benefit of lowering the performance cost for every violation when the feature is not used.

I added a test, as I believe the behavior should be guaranteed from now on.

### CODE REVIEW

**Comment by romaricdrigon** [comment]
I just realized my fix introduces another issue - which is that `{{ label }}` could be only in the translated form label.
Hopefully there is a way around, which is that if `label` is `null`, it will never contain `{{ label }}` - so we can keep the optimization and BC for form classes.
I'm fixing my PR.

**Review by nicolas-grekas — APPROVED** [review]
/cc @a-menshchikov FYI

File: src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php [hunk 1]
```diff
@@ -131,44 +131,46 @@ public function mapViolation(ConstraintViolation $violation, FormInterface $form
...
+                            $scope->getName(),
+                            (string) $scope->getPropertyPath(),
+                        ],
+                        $labelFormat
+                    );
+                } else {
+                    $label = $scope->getConfig()->getOption('label');
+                }
+
+                if (false !== $label) {
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php:156]: Why is this necessary? If `$label` is `null`, it will never be `false`. So the previous code was already working as expected, wasn't it?

File: src/Symfony/Component/Form/Tests/Extension/Validator/ViolationMapper/ViolationMapperTest.php [hunk 2]
```diff
@@ -1738,4 +1738,27 @@ public function testMessageWithLabelFormat2()
...
+
+    public function testTranslatorNotCalledWithoutLabel()
+    {
+        $renderer = $this->getMockBuilder(FormRenderer::class)
+            ->setMethods(null)
+            ->disableOriginalConstructor()
+            ->getMock()
+        ;
+        $translator = $this->getMockBuilder(TranslatorInterface::class)->getMock();
+        $translator->expects($this->any())->method('trans')->willReturnCallback(function () {
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Tests/Extension/Validator/ViolationMapper/ViolationMapperTest.php]:
```php
$translator = $this->createMock(TranslatorInterface::class);
$translator->expects($this->never())->method('trans');
```

What about this instead?

File: src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php (hunk 1 above)
Comment (romaricdrigon) [reply src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php:156]: Yes, both versions are giving the same result. But moving up the `false !== $label` test, instead of leaving it around line 162, avoid an extra an extra test (`null === $label`).

File: src/Symfony/Component/Form/Tests/Extension/Validator/ViolationMapper/ViolationMapperTest.php (hunk 2 above)
Comment (romaricdrigon) [reply src/Symfony/Component/Form/Tests/Extension/Validator/ViolationMapper/ViolationMapperTest.php]: Sure! I'm not supra familiar with PHPUnit mocks, your version is quite better.

**Comment by romaricdrigon** [comment]
(rebased branch, to try get CI green)

**Comment by fabpot** [comment]
We've just moved away from `master` as the main branch to use `5.x` instead. Unfortunately, I cannot reopen the PR and change the target branch to `5.x`. Can you open a new PR referencing this one to not loose the discussion? Thank you for your understanding and for your help.

**Comment by romaricdrigon** [comment]
Done
